### DISCUSSION TITLE
Handling external delegate

### DISCUSSION THREAD

**Discussion by JeromeTonnelierOgury** [pointfreeco/swift-composable-architecture#2346] [General]
Hi again!

I have an external SDK (let's say a VideoStreamer) and I want to handle it from my reducer.
Basically I can load a video and I get the various callbacks from a classic delegate.

I guess my "main" Recuder should have a copy of this VideoStreamer (it's a class, to by reference should be fine), but how can I get back some stuff inside the Reducer from a delegate from the VideoStreamer ? 
I initially thought of a Combine object or an async variable, but it seems that many people use an AsyncStream. What is the best approach to handle a delegate ? 

Thanks!

**Comment by mbrandonw** [comment]
Hi @JeromeTonnelierOgury, all external SDK's should be considered dependencies and so should be wrapped in an interface and registered with the dependency management system that comes with the library. You should not hold any reference types in your state because that makes equatability essentially impossible.

The dependency interface you create will expose endpoints for the various things you can invoke on `VideoStreamer`, and the dependency can feed data back into the system via either a publisher or async sequence. While async sequences are the new hotness in Swift, they also have their drawbacks, so it really just depends whichever you are more comfortable with. 

We have lots of examples of implementing this kind of dependency in our demos and case studies:

* [Speech recognizer](https://github.com/pointfreeco/swift-composable-architecture/blob/main/Examples/Standups/Standups/Dependencies/SpeechRecognizer.swift) from the Standups app
* [Audio player](https://github.com/pointfreeco/swift-composable-architecture/tree/main/Examples/VoiceMemos/VoiceMemos/AudioPlayerClient) and [Audio recorder](https://github.com/pointfreeco/swift-composable-architecture/tree/main/Examples/VoiceMemos/VoiceMemos/AudioRecorderClient) from voice memos app
* And _lots_ of examples in our [isowords](https://github.com/pointfreeco/isowords/tree/main/Sources) app.

**Reply by JeromeTonnelierOgury** [reply]
Thanks a lot for all of this, I'll have a look then!

**Comment by JeromeTonnelierOgury** [comment]
If I get it right @mbrandonw, I must do something like this ? 
```swift
struct TCAVideoPlayer {
  // sasically transform every delegate method into a @Sendable var
   var loaded: @Sendable () -> async Bool
}

extension TCAVideoPlayer: DependencyKey {
   static let liveValue: TCAVideoPlayer {
      let originalPlayer = VideoStreamer()
      return TCAVideoPlayer(
         loaded: { //implementation }
   }
}
```

Is that the right direction ? Thanks 🙏

**Reply by mbrandonw** [reply]
That is definitely a good start. Things may get more complicated as you add more endpoints and functionality, but you've got the basics so far. Feel free to reply back in this thread if you run into a complication in the future.

Oh also, be sure to extend `DependencyValues` with a computed property so that you can actually access the dependency via the `@Dependency` property wrapper.

**Reply by JeromeTonnelierOgury** [reply]
Ok Thanks a lot, I 'll see how it goes from here :-)

**Reply by JeromeTonnelierOgury** [reply]
Oh quick one though : my VideoStreamer is actually a protocol with generic type. How can I provide an effective implementation for various context ? 
Can I extend TCAVideoPlayer with a Generic like ` TCAVideoPlayer<T: VideoStreamer>' and after, inject real time dependencies with `TCAVideoPlayer<VideoStreamerLocal>`and `TCAVideoPlayer<YoutubeVideoStreamer`for instance ? Thanks a lot for your help <3

**Reply by mbrandonw** [reply]
Can you describe more about what the generic on `VideoStreamer` represents? Or can you share the protocol and a sample conformance?

**Reply by JeromeTonnelierOgury** [reply]
Yes basically, let's say I have VideoManagers for various sources (let's say Local file, Youtube and Dailymotion). All VideoManagers would extend a basic procotol
```swift
protocol VideoManager {
   associatedtype Video // the video object to play, loaded from an external SDK
   var video: Video { get }
   func load()
   func play()
   var videoDelegate { get set } // here is the delegate for the load and play operation
}
```
and then I can construct let's say a YoutubeVideoManager
```swift
struct YoutubeVideoManager: VideoManager {
   typealias Video = YoutubeVideo
   var video: YoutubeVideo = YoutubeVideo()
   // add the various stuff
   func load() {
      video.delegate = self
      video.load()
   }
}

extension YoutubeVideoManager: YoutubeVideoHandler {
   func youtubeVideoDidLoad() {
      delegate.didLoad()
   )

   func youtubeVideoDidFailToLoad(_ error: Error) {
      delegate.didFail(with: error)
   )
}
```
And I have dedicated managers for every video type. This part I cannot change of course

Now I construct a SwiftUI View that will manage Videos through a VideoManager object (non specialized) and I have to pass to the associated Reducer an actual implementation of a VideoManager

I hope it helps and it is clearer! Let me know othewise!
Thanks

EDIT: to ease my TCA integration, I'm thinking that I will move the delegate to a Combine object with PassthroughSubject for every delegate method. I guess it will be easier to integrate with TCA in the end right ?

**Reply by JeromeTonnelierOgury** [reply]
@mbrandonw do you have enough information, or do you want me to expand it a bit ? Thanks 😊

**Reply by Alex293** [reply]
Why not make 3 separate dependencies ? I don’t think you will be able to have associated type protocol without erasing in the dependencies

**Reply by JeromeTonnelierOgury** [reply]
Thanks for your answer @Alex293. Actually, I have 4 different implementations for this associated type, and I could have more (hence the associatedtype ^^ ).
If I have to create a new dependency for each new implementation, that means that the main reducer would have to know which dependency to use depending on the associatedtype, which is counter productive right ?

**Reply by Alex293** [reply]
I thought you wanted to use all implementations separately. If you only want one dependency you need to erase it but you won't be able to use the associated type.

**Reply by JeromeTonnelierOgury** [reply]
I don't think that's a problem since all generic implementation have the same protocol requirement, so ITE, I only use the protocol stuff, I don't need external access to the associated types

**Reply by JeromeTonnelierOgury** [reply]
How can I achieve some type erasure and inject the actual good implementation into the TCA reducer ?

**Reply by Alex293** [reply]
Well it's quite hard to get the full picture from the protocols above :)

The erasure I was suggesting is making a struct AnyVideoPlayer without any generics which you can construct with your different players and only expose non generic api. Something like :
```
struct AnyPlayer {
    private let _play: () -> Void

    init(player: some Player) {
        self._play = { player.play()  }
    }
    func play() { _play() }
}
```
but you can achieve the same thing if you remove then associated type and any generics to your protocol. What would you do anyway with the `video` property ?

Another approach you be to wrap into an enum your different Video kinds and unifying your protocols.

**Reply by JeromeTonnelierOgury** [reply]
I guess I can create some dedicated dependencies, but when I'll construct the View, and pass it a Store, How can I say to that store "use this dependecy" ? Is there a way, at runtime too, to inject a dependency inside a Reducer ?

**Reply by mbrandonw** [reply]
Hi @JeromeTonnelierOgury, I see you posted this question on the Swift forums too, and I just want to mention that I don't really think this is a TCA-specific question anymore. 

Abstractly, you have a generic dependency/protocol that you want to inject into your feature. You really have no choice but to either introduce a generic to the feature that touches the dependency (which is viral and will cause generics to be introduced to any feature touching this feature), or introduce some kind of erasure like @Alex293 mentioned. The latter is most likely the way to go, and so you should focus your energy on figuring out how you can erase your protocol in the most reasonable way possible.

This problem exists whether you are using TCA or not, and so you may find that asking the question in a more general forum, such as the [Apple developer forums](https://developer.apple.com/forums/) will get you more attention.

**Reply by JeromeTonnelierOgury** [reply]
Thanks @mbrandonw but I have to disagree if you may 😊 

Yes, it's an issue with generics and erasure, but my issue here is : how can I use the `@Dependency` TCA feature with a generic in the end. 
If it was not for the `@Dependency`, I would surely know how to inject it in any object. I would just have to declare a `var videoManager: VideoManager` and that's it. 

But here, I cannot use a variable as you stated, I must use a `@Dependency` along with everything else (the `@sendable` wrapper and so on)
No one ever encountered this issue with injecting a generic dependency ? 
I guess it would look something like this, maybe this can help you see where I am going ? 
```swift
struct TCAVideoPlayer {
   var videoManager: VideoManager // base protocol object, erased on can say ? 
  // basically transform every delegate method into a @Sendable var
   var loaded: @Sendable () -> async Bool
}

extension TCAVideoPlayer: DependencyKey {
   static func liveValue(_: manager VideoManager): TCAVideoPlayer {
      let originalPlayer = VideoStreamer()
      return TCAVideoPlayer(
         videoManager: manager,
         loaded: { //implementation }
   }
}

// and then later on
// this would specialize the overall behavior with a YoutubeVideoPlayer
let videoView = VideoView(store: VideoStore(withDependency: TCAVideoPlayer.liveValue(YoutubeVideoPlayer)))
```

I don't know if it's clearer or not ? 
Thanks for your help, much appreciated 🙏

**Reply by Alex293** [reply]
Can you provide a trimmed down version of your view ? 
Is it generic ? If your view is generic then you cannot use erasure and need multiple dependencies, if your view isn't I believe you should remove any associated object / generic from you dependency and keep them at the implementation level

**Reply by JeromeTonnelierOgury** [reply]
I don't have the View w/o TCA anymore, but basically it would look like this (this would be for the VideoBar which would have the 2 buttons `load`and `show`

```swift
struct VideoBar: View {
    let videoManager: VideoManager
    var body: some View {
            HStack(spacing: 10) {
                Group {
                    Button("Load") {
                        videoManager.load()
                    }
                    Button("Show") {
                        videoManager.show()
                    }
                }
                .tint(.blue)
                .foregroundColor(.black)
                .buttonBorderShape(.capsule)
                .buttonStyle(.borderedProminent)
            }
        }
}
```

and I would instanciate it with a simple `let view = VideoBar(videoManager: YoutubeVideoManager)` when I want to handle youtube videos and  `let view = VideoBar(videoManager: LocalVideoManager)` when I want to handle local file.

But I can't just use the same for the store as it seems that I have to use a dependency rather than a simple `var`, that's the whole point. In the end, the VideoManager protocol acts as an erasure since I don't need to know what's going on inside each method. I just need access to `load` and `show`, do you see what I mean ?

**Reply by mbrandonw** [reply]
I'm not sure I see why you can't use `VideoManager` as a dependency in the exact same way it is provided to the view you demonstrate above. You would do something like:

```swift
private enum VideoManagerKey: TestDependencyKey {
  static let testValue: any VideoManager = /* Some kind of mock or unimplemented conformance */
}
extension DependencyValues {
  var videoManager: any VideoManager {
    get { self[VideoManagerKey.self] }
    set { self[VideoManagerKey.self] = newValue }
  }
}
```

That provides a type erased `VideoManager` to anyone using `@Dependency(\.videoManager)` just as you did in your view above.

It's worth noting that in this erasure you can't really do anything with the `video` property. It's effectively an `Any`. But it doesn't seem you are using the `video` property in the view. Maybe you can even drop that associated type.
